File: lua/pythagorean-triplet/pythagorean-triplet.lua

local function is_triplet(...)
	local triplet = table.pack(...)
	table.sort(triplet)
	return triplet[1]^2 + triplet[2]^2 == triplet[3]^2
end

local function triplets_with(range)
	local max = range.max_factor
	local min = range.min_factor or 0
	local result ={}
	for i = min, max do
		for j = i+1, max do
			for k = j+1, max do
				if i^2 + j^2 == k^2 then
					table.insert(result, {i,j,k})
				end
			end
		end
	end
	if range.sum then
		for i= #result, 1, -1 do
			if result[i][1] + result[i][2] + result[i][3] ~= range.sum then 
				table.remove(result, i)
			end
		end
	end
	table.sort( result, function(a,b) return a[1] < b[1] end)
	return result
end

return {
	is_triplet = is_triplet,
	triplets_with =triplets_with
}

File: javascript/pythagorean-triplet/pythagorean-triplet.js

export class Triplet {
  constructor(s) {
    this.sides_ = Object.values(arguments);
  }

  sum() {
    return this.sides_.reduce((sum, ele) => sum+ ele, 0);
  }

  product() {
    return this.sides_.reduce((sum, ele) => sum* ele, 1);
  }

  isPythagorean() {
    return this.sides_[0]**2 +this.sides_[1]**2 ===this.sides_[2]**2 ;
  }

  static where(options) {
    let min = options.minFactor || 1;
    let max = options.maxFactor ;
    let sum = options.sum;
    let result =[];
    for(let c= min; c <= max; ++c){
      for(let b= min; b < c; ++b){
        let a = Math.sqrt(c**2 - b**2);
        if(Number.isInteger(a) && (a <b) && (a >= min)){
          let tempP = new Triplet(a,b,c);
          if((tempP.sum() === sum || !sum)){
            result.push(tempP);
          }
        }
      }
    }
    return result;
  }
}
